app/src/main/java/unipd/se18/ocrcamera/inci/TextAutoCorrection.java

```diff
@@ -0,0 +1,164 @@
+
+        //search all words composed by alphanumeric or hyphen characters
+        Pattern pattern = Pattern.compile("[a-zA-Z0-9-]+");
+        Matcher matcher = pattern.matcher(text);
+
+        //try to correct each word
+        int findFromIndex = 0; //index after which we look for the next match
+        while (matcher.find(findFromIndex)) {
+            
+            String word = matcher.group();
+            findFromIndex = matcher.end();
+
+            if(word.length()>=minChars) {
+                String corrected = correctWord(word);
+                if (!corrected.equals(word)) {
+
+                    Log.d(TAG, "word " + word + " corrected with " + corrected);
+
+                    //substitute with the word found
+                    text = text.substring(0, matcher.start()) + corrected + text.substring(matcher.end());
+
+                    //take into account difference in length between original and corrected word
+                    if(corrected.length() != word.length()) {
+                        matcher = pattern.matcher(text);
+                        findFromIndex += corrected.length() - word.length();
+                    }
+                }
+            }
+
+        }
+
+        return text;
+    }
+
+
+    /**
+     * format the text in order to increase the probability to match ingredients in the INCI DB
+     * @param text Text to be formatted
+     * @return Text formatted
+     */
+    private String formatText(String text){
+        //merge the word before and after hyphen + new line (e.g. "ceta- \n ril" into "cetaryl")
+        text = text.replaceAll(" *- *[\\n\\r]+ *", "");
+
+        //ignoring case by converting all into upper case
+        text = text.toUpperCase();
+        return text;
+    }
+
+    /**
+     * Correct a single word by searching for the most similar in word list
+     * @param word The word to be corrected
+     * @return Best candidate word from word list. If no words within maxDistance is found, the same word is returned.
+     */
+    private String correctWord(String word){
+
+        //percentage distance below which we substitute the word with the term found in dictionary
+        // (during testing i found out that above 25% the confidence does not improve by much and
+        // also increases chance of correcting words not related to ingredients)
+        final double maxDistance = 0.25;
+
+        //Searches the tree for elements whose distance satisfy maxDistance
+        Set<BkTreeSearcher.Match<? extends String>> matches =
+                searcher.search(word, (int) (word.length()*maxDistance));
+
+        //find the word with minimum distance
+        int minDistance = Integer.MAX_VALUE;
+        String closest = "";
+        for (BkTreeSearcher.Match<? extends String> match : matches){
+            if(match.getDistance() < minDistance) {
+                minDistance = match.getDistance();
+                closest = match.getMatch();
+            }
+        }
```

Comment:
In this for cycle instead of compare getDistance with minDistance and then calculate new minDistancek, could be better first calculate new minDistance and then compare it. So u dont have to set min Distance to MAX_VALUE before the cicle

Reply:
I didn't quite understand sorry, could you be more specific?